src/test/java/com/android/volley/toolbox/BasicAsyncNetworkTest.java

```diff
@@ -0,0 +1,509 @@
+            httpNetwork.setBlockingExecutor(executor);
+            Request<String> request = buildRequest();
+            request.setRetryPolicy(mMockRetryPolicy);
+            doThrow(new VolleyError()).when(mMockRetryPolicy).retry(any(VolleyError.class));
+            httpNetwork.performRequest(request, mockCallback);
+            verify(mockCallback, times(1)).onError(any(VolleyError.class));
+            verify(mockCallback, never()).onSuccess(any(NetworkResponse.class));
+            // should not retry any 500 error w/ HTTP 500 retries turned off (the default).
+            verify(mMockRetryPolicy, never()).retry(any(VolleyError.class));
+            reset(mMockRetryPolicy, mockCallback);
+        }
+    }
+
+    @Test
+    public void notModifiedShortCircuit() throws Exception {
+        MockAsyncStack mockAsyncStack = new MockAsyncStack();
+        List<Header> headers = new ArrayList<>();
+        headers.add(new Header("ServerKeyA", "ServerValueA"));
+        headers.add(new Header("ServerKeyB", "ServerValueB"));
+        headers.add(new Header("SharedKey", "ServerValueShared"));
+        headers.add(new Header("sharedcaseinsensitivekey", "ServerValueShared1"));
+        headers.add(new Header("SharedCaseInsensitiveKey", "ServerValueShared2"));
+        HttpResponse fakeResponse = new HttpResponse(HttpURLConnection.HTTP_NOT_MODIFIED, headers);
+        mockAsyncStack.setResponseToReturn(fakeResponse);
+        BasicAsyncNetwork httpNetwork = new BasicAsyncNetwork.Builder(mockAsyncStack).build();
+        httpNetwork.setBlockingExecutor(executor);
+        Request<String> request = buildRequest();
+        httpNetwork.performRequest(request, mockCallback);
+        verify(mockCallback, times(1)).onSuccess(any(NetworkResponse.class));
+        verify(mockCallback, never()).onError(any(VolleyError.class));
+        reset(mMockRetryPolicy, mockCallback);
+    }
+
+    @Test
+    public void performRequestSuccess() throws Exception {
+        MockAsyncStack mockAsyncStack = new MockAsyncStack();
+        HttpResponse fakeResponse =
+                new HttpResponse(
+                        200,
+                        Collections.<Header>emptyList(),
+                        "foobar".getBytes(StandardCharsets.UTF_8));
+        mockAsyncStack.setResponseToReturn(fakeResponse);
+        BasicAsyncNetwork httpNetwork = new BasicAsyncNetwork.Builder(mockAsyncStack).build();
+        httpNetwork.setBlockingExecutor(executor);
+        Request<String> request = buildRequest();
+        Entry entry = new Entry();
+        entry.etag = "foobar";
+        entry.lastModified = 1503102002000L;
+        request.setCacheEntry(entry);
+        httpNetwork.performRequest(request, mockCallback);
+        verify(mockCallback, times(1)).onSuccess(any(NetworkResponse.class));
+        verify(mockCallback, never()).onError(any(VolleyError.class));
+        reset(mMockRetryPolicy, mockCallback);
+    }
+
+    @Test(expected = IllegalStateException.class)
```

Comment:
Can we use Assert.assertThrows without adding a new dependency? If so, that's preferable since it more tightly describes the code expected to fail, rather than just letting any IllegalStateException thrown anywhere inside this test cause the test to pass.

Reply:
I don't think so, since that's only in JUnit5. I'm not aware of another way to test that that specific code causes it.